### PR TITLE
feat: add local_window_size attribute to Attention op (opset 25)

### LINKED ISSUE #7914: Add local_window_size attribute to Attention op for sliding window support
## Motivation

Modern LLMs like Gemma4, Mistral, and others use hybrid attention with both full (global) attention and sliding window (local) attention layers. The sliding window pattern only attends to the most recent N tokens, reducing memory and compute for long sequences.

The ONNX `Attention` op (opset 23+) currently has no mechanism to specify a local attention window. The `com.microsoft.GroupQueryAttention` contrib op supports this via `local_window_size`, but the standard ONNX `Attention` op does not.

## Problem

Without `local_window_size`, models with sliding window attention cannot use static (pre-allocated) KV caches with TensorScatter for inplace updates. The static KV buffer retains all historical entries, and without a window parameter, the Attention op incorrectly attends to stale entries beyond the window.

This forces sliding window layers to use either:
- **Dynamic KV caches** (concat + slice, less efficient, prevents CUDA graph capture)
- **Vendor-specific contrib ops** (e.g., `com.microsoft.GroupQueryAttention`)
- **Explicit attention bias construction** (materialize a full causal+window mask, memory-intensive)

## Proposal

Add an optional `local_window_size` integer attribute to the ONNX `Attention` op:

- **When set to a positive value**: The op only attends to the most recent `local_window_size` key/value positions. Positions outside the window are masked (treated as -inf before softmax).
- **When not set or set to -1** (default): Full causal attention is used. This is fully backward compatible.
- **Interaction with `is_causal`**: When both `is_causal=1` and `local_window_size > 0`, the op applies a causal sliding window mask (each query attends to at most `local_window_size` preceding keys).

This enables efficient sliding window attention with static KV caches across all ONNX runtimes, without requiring vendor-specific ops.

## Use Cases

- **Gemma4**: Alternating sliding (window=512/1024) and full attention layers within the same model
- **Mistral / Mixtral**: Sliding window attention (window=4096)
- **Qwen3.5**: Hybrid attention with mixed local/global layers
- **Any hybrid attention architecture** combining local and global attention patterns

## Prior Art

- `com.microsoft.GroupQueryAttention` already supports `local_window_size` as an integer attribute with the same semantics proposed here
- PyTorch's `F.scaled_dot_product_attention` does not have a built-in window parameter but frameworks like FlashAttention support it via custom masks
- The ONNX `Attention` op's existing `is_causal` attribute demonstrates that masking behavior can be controlled via simple integer attributes

### PR DESCRIPTION
**Pull request by FuZoe**
### Motivation and Context

Fixes https://github.com/onnx/onnx/issues/7914

Modern LLMs such as Gemma4, Mistral, and Qwen3.5 use sliding window attention to limit the context each token can attend to, enabling efficient inference with static KV caches and reducing the quadratic attention cost. This PR adds an optional `local_window_size` attribute to the Attention operator (bumped to opset 25) to support this pattern.

### Changes

**Updating existing operator (per [AddNewOp.md checklist](https://github.com/onnx/onnx/wiki/Checklist-for-updating-an-existing-operator)):**

| Step | Status |
|------|--------|
| Move current schema to `old.cc` | Done — Attention ver24 moved to `onnx/defs/nn/old.cc` |
| Create new version in `defs.cc` | Done — Attention ver25 with `local_window_size` attribute (default: -1) |
| Update operator set registration | Done — `onnx/defs/operator_sets.h` |
| Add interned symbol | Done — `onnx/common/interned_strings.h` |
| Version converter adapter | Done — `onnx/version_converter/adapters/Attention_25_24.h`, compatible upgrade (24->25), custom downgrade (25->24, rejects if `local_window_size > 0`) |
| Update reference implementation | Done — `onnx/reference/ops/op_attention.py`, added `_apply_sliding_window()` helper |
| Function body builder | Done — sliding window mask logic in `defs.cc` using Range/Sub/Less/Where ops |
| Node tests | Done — 5 new tests: `local_window`, `local_window_causal`, `local_window_with_attn_mask`, `local_window_default`, `local_window_omitted` |
| Upgrade/downgrade tests | Done — 2 upgrade tests (24->25), 2 downgrade tests (25->24 success + rejection) |
| Regenerate docs | Done — `Operators.md`, `Changelog.md`, `TestCoverage.md` (69 Attention test cases) |

**Semantics:**

When `local_window_size` is set to a positive value, positions where `(past_sequence_length + query_index) - key_index >= local_window_size` are masked with `-inf` before softmax. This limits how far back each query can attend. With `is_causal=1`, each query attends to at most `local_window_size` keys including itself. Future positions remain visible unless `is_causal=1` is also set. Default value is `-1` (full attention, fully backward compatible).

### CODE REVIEW

File: onnx/reference/ops/op_attention.py [hunk 1]
```diff
@@ -43,6 +43,24 @@ def _apply_causal(mask, past_sequence_length):
...
+    Future positions (where j > past_sequence_length + i) remain visible;
+    use ``_apply_causal`` to also mask those.
+    The modification is done inplace.
+    """
+    q_sequence_length, total_sequence_length = mask.shape[-2:]
+    row_idx = np.arange(q_sequence_length).reshape(-1, 1) + past_sequence_length
+    col_idx = np.arange(total_sequence_length).reshape(1, -1)
+    diff = row_idx - col_idx
+    window_mask = diff >= local_window_size
+    mask[..., window_mask] = -np.inf
```
[the remark this answers is not in the document]
Comment (FuZoe) [reply onnx/reference/ops/op_attention.py]: `mask[..., window_mask]` works correctly for any rank. The ellipsis broadcasts the 2D boolean index over all leading dims (batch, heads). Same pattern as `_apply_causal` at line 42. No change needed.

**Comment by FuZoe** [comment]
The merge conflicts are resolved now. Please continue the review when convenient.

**Comment by titaiwangms** [comment]
## Review: `local_window_size` on Attention (opset 25)

Thanks for this — sliding-window attention is a real, well-motivated need (Gemma/Mistral/Qwen), and the overall shape is right: adding it to `Attention` as an additive-mask attribute (rather than a new op or a caller-materialized float mask) lets backends pattern-match to fused sliding-window kernels, and the **causal** case is defined correctly. My comments are one spec-level decision, one correctness bug, and two interop/CI items.

### Normative basis I'm reviewing against
To keep this from being "align with one runtime", I'm treating the source of truth as **(a) the masking math + this op's function-body decomposition**, with **HF Transformers, PyTorch FlexAttention, and ORT as three independent witnesses** of the established sliding-window semantics. Where all three agree, that's the spec; anything specific to a single runtime's kernel is an implementation detail and should stay out of the spec.

### Why the breadth gap vs any kernel is fine
ONNX `Attention` is intentionally a superset of any fused kernel (it covers MHA/GQA/MQA, self/cross-attention, etc.), and it ships a function-body decomposition — so any conforming runtime can *execute* every spec'd combination even without a fused path. So I'm **not** concerned that the spec covers combinations a given kernel doesn't fuse. What must hold is that on the **intersection** of configs the kernels do claim (causal decode, GQA-shaped, supported dtypes), the ONNX reference and those runtimes are **bit-identical**. The items below close exactly that intersection, plus one under-motivated corner.

---

### 🔴 Blocking

**B1 — The `is_causal=0` window mode doesn't match the cross-framework definition of a sliding window.**
A sliding window of size `W` is, across the ecosystem, intrinsically causal — the upper bound `key_idx ≤ query_idx` is part of the definition:
- HF Transformers (Mistral): query `i` attends `[i−W+1, i]`; `sliding_window=None` ⇒ full causal. (`src/transformers/models/mistral/modeling_mistral.py`, sliding-window mask construction.)
- PyTorch FlexAttention canonical sliding window: `mask = (kv_idx >= q_idx-(W-1)) & (kv_idx <= q_idx)` — note the `kv_idx <= q_idx` term.
- ORT `GroupQueryAttention`: "left_window_size for local attention (like Mistral)", always causal. (`docs/ContribOperators.md` GroupQueryAttention; kernel `onnxruntime/contrib_ops/cpu/bert/gqa_attention_base.h`, softmax/mask loop.)

This PR's `is_causal=0 && local_window_size>0` drops that upper bound: a query attends `[p−W+1, ∞)` — only the far past is masked, all future stays visible. That is neither the causal sliding window above, nor the symmetric/bidirectional local window of Longformer/BigBird (`|i−j| < W`). It's a one-sided lower bound that no mainstream framework defines, and all three models cited in the PR description use the causal form. **Recommendation:** make `local_window_size` part of the causal frontier (i.e. it implies the causal upper bound). If a bidirectional local window is ever wanted, add it separately and make it symmetric. This is schema-visible, so worth an op-owner sign-off rather than settling in review.

**B2 — Sliding window vs the two KV-cache modes: composes with dynamic cache, but not static cache (correctness + a per-batch dimensionality gap worth discussing).**

*Dynamic cache (`past_key`/`past_value`, internal growing cache):* already correct. The offset is the scalar `past_key.shape[2]`, which is exactly what the window uses (`WinRangeRowPast = WinRangeRow2D + PastKVSeqLen`; reference `past_sequence_length = past_key.shape[2]`). Chunked prefill (`q>1`, past>0) is fine too. ✅

*Static cache (`nonpad_kv_seqlen`, no `past_key`, preallocated buffer):* broken, and the fix is more than a constant swap:
1. **Wrong offset.** With no `past_key` the window anchors at `0`, but the per-batch frontier is `offset[b] = nonpad_kv_seqlen[b] − q_sequence_length` (the same offset #8068 introduces for `is_causal`). Today stale, out-of-window keys leak in.
2. **The window mask must become per-batch (4D).** In static cache the offset *varies per batch*, so the window position does too — but the PR currently builds a batch-independent 2D `[Sq, Skv]` mask. It has to broadcast a per-batch offset and become `[batch, 1, Sq, Skv]`, exactly as `_apply_causal` already does. So this is a structural change to the window sub-graph, not just wiring in `CausalOffsetPerBatch`.

Composition itself is fine once the above hold: causal + window + padding are all additive `-inf`, so they compose order-independently to the correct set `(qpos−W, qpos] ∩ [0, qpos] ∩ [0, nonpad[b])`.

*Out of scope but worth flagging for discussion:* a **circular/ring-buffer** static cache (e.g. TensorScatter's `"circular"` mode) has non-contiguous valid keys (wrap-around modulo `max_sequence_length`). Both the causal frontier and this window assume a contiguous valid prefix `[0, nonpad)` via `Range`, so **neither** handles ring buffers — a pre-existing Attention limitation the window inherits, not one it introduces. If sliding-window + ring buffer is an intended target, it needs a separate design.

*Illustration of the offset bug* — static cache, `total=6, q=2, nonpad=4, is_causal=1, W=2`: query row 0 (abs pos 2) should attend `{1,2}`, but with `offset=0` the window masks nothing and it attends `{0,1,2}` (stale key 0 leaks).

Concretely this PR **depends on #8068**: land it first, rebase, then make the window per-batch on the same offset.

**B3 — `local_window_size = 0` meaning "full attention" is internally inconsistent and matches no framework.**
Two arguments, neither leaning on a single backend:
1. *Self-consistency:* it's a window **size**, so semantics should be monotonic — `W=1`⇒1 key, `W=2`⇒2 keys, … and `W→∞`⇒full. Under that reading `0` is the *smallest* window (empty / 0 keys), yet this PR makes `0` mean the *largest* (unbounded) — the one value whose meaning is inverted. Least-surprise violation regardless of runtime.
2. *Prevailing convention:* every witness disables via a distinguished *absence*, never `0` — HF `sliding_window=None`, PyTorch "no mask_mod", ORT `-1`.

Please reserve an out-of-domain sentinel for "disabled" (`-1`, already the default) and either **reject `0`** in the schema or define it as the empty window — not full attention.

**B4 — Missing committed backend test-data artifacts (CI).**
The 5 new `export_attention_local_window*` cases are added to `attention.py` and counted in `TestCoverage.md`, but no `onnx/backend/test/data/node/test_attention_4d_local_window*/` directories are in the diff. These dirs are tracked and CI expects them — please regenerate and commit (cf. #8068, which regenerated the full set in one commit).

### ✅ Confirmed correct (no change needed)
- The causal off-by-one (`W` keys including self, `[p−W+1, p]`) matches the cross-framework definition — HF Mistral, PyTorch FlexAttention, and ORT GQA all agree.
- C++ function body and Python reference describe identical masked sets (`Less(diff, W)` is the exact complement of `diff ≥ W`, boundary included).
- Fully-masked rows are absorbed by the pre-existing zero-row guard → no new NaN path.

---

### 🟡 Non-blocking
- **Test coverage:** add window + `past_key` and window + `is_causal` + `nonpad_kv_seqlen` cases — the latter would have surfaced B2.
- **Doc the relationship:** state how this relates to the causal-only sliding window in HF/PyTorch/ORT, to prevent silent migration mismatches.
- **Float-mask dtype:** `Add(AttnBiasCausalOrNot, WinMask)` mixes a float-typed mask with a possibly fp16/bf16 bias; this mirrors a *pre-existing* pattern (causal `Add` in `utils.cc`), so not introduced here, but a `CastLike` would be cheap op-wide hardening.
- **Readability:** `WinRangeRowPast` actually holds absolute query positions (e.g. `WinAbsQueryPos`); `_apply_sliding_window` docstring uses undefined `i`/`j` (use `query_index`/`key_index`); `Win`/`Window` constant-prefix inconsistency; "preceding keys" vs "including itself" wording conflict.

### Bottom line
Design and the causal semantics are sound, and the residual breadth gap vs any single kernel is the healthy kind (covered by decomposition). Blockers: decide the non-causal mode (B1), fix the static-cache offset + per-batch window on top of #8068 (B2), align the `0`/`-1` sentinel (B3), commit test data (B4). With those, ONNX matches HF/PyTorch/ORT bit-for-bit on the overlap, and everything else degrades gracefully via the function body.

**Comment by FuZoe** [comment]
Thanks for the detailed review and the cross-framework normative basis — it's exactly the kind of rigor this spec-level change needs.

### B1 — Window implies causal

I initially kept `is_causal=0 + local_window_size > 0` as a general-purpose lower-bound primitive, thinking it might serve future bidirectional-local patterns. But looking at it more carefully: (1) none of the three models motivating this PR use that mode, (2) the resulting semantics — mask only the far past, leave all future visible — doesn't correspond to any established attention pattern (it's neither Longformer's symmetric `|i−j| < W` nor the standard causal window), and (3) specifying an unmotivated corner case weakens the spec by creating a combination that no backend will optimize for and no model will use.

I'll make `local_window_size > 0` imply the causal upper bound unconditionally. If a symmetric bidirectional window is needed later, it deserves its own attribute with explicit symmetric semantics rather than overloading this one.

### B2 — Static cache per-batch window (on top of #8068)

I've read through #8068 — the core insight is introducing `offset[b] = nonpad_kv_seqlen[b] − q_sequence_length` for bottom-right causal alignment on the external-cache path. My window mask needs to build on that same per-batch offset:

- The window condition becomes `key_idx < (query_in_block_idx + offset[b]) − W + 1`, which requires broadcasting over the batch dimension
- The current 2D `[Sq, Skv]` mask must become `[batch, 1, Sq, Skv]` to accommodate per-batch `offset[b]`
- The function body will use `Unsqueeze` + the same `CausalOffsetPerBatch` sub-graph that #8068 introduces

I'll rebase once #8068 lands and build the per-batch window on that foundation.

### B3 — Sentinel semantics for `0`

A window *size* should be monotonic in coverage: `W=1` → 1 key (self only), `W=2` → 2 keys, …, `W→∞` → full. Having `0` mean "unbounded" inverts this monotonicity and contradicts the prevailing convention (HF `sliding_window=None`, ORT `-1`, PyTorch "no mask_mod").

I'll make `-1` the sole "disabled" sentinel (already the default), and add a schema constraint rejecting `0` with a clear error message directing users to use `-1` for full attention.

### B4 — Backend test data

Will regenerate the `onnx/backend/test/data/node/test_attention_4d_local_window*/` directories and commit them. I noticed #8068 regenerated the full set in one commit — I'll follow the same approach.

### Non-blocking items

- **Test coverage**: Will add window + `past_key` (dynamic cache) and window + `is_causal` + `nonpad_kv_seqlen` (static cache) cases — the latter would have caught B2 earlier.
- **Naming**: `WinRangeRowPast` → `WinAbsQueryPos` (it holds absolute query positions, not a "range"); unify `Win`/`Window` prefix; replace `i`/`j` in docstrings with `query_index`/`key_index`.
- **Documentation**: Will add a note clarifying the relationship to HF/PyTorch/ORT causal sliding window semantics.

Holding off on code changes until #8068 merges. Will update the PR once rebased.

**Comment by FuZoe** [comment]
Closing this PR in favor of #8108, which is a clean rebase onto the latest main (post-#8068). All review feedback has been addressed — implicit causal semantics, per-batch offset broadcasting, invalid value rejection, and test data generation.
